core/src/main/java/com/linecorp/armeria/internal/common/RequestContextUtil.java

```diff
@@ -48,13 +55,45 @@
     private static final Set<Thread> REPORTED_THREADS =
             Collections.newSetFromMap(new MapMaker().weakKeys().makeMap());
 
+    private static final ContextStorage contextStorage;
+
+    static {
+        final List<ContextStorageProvider> providers = Streams.stream(
+                ServiceLoader.load(ContextStorageProvider.class)).collect(toImmutableList());
+
+        if (providers.isEmpty()) {
+            contextStorage = new ContextStorageThreadLocal();
+        } else {
+            final ContextStorageProvider provider = providers.get(0);
+            if (providers.size() > 1) {
```

Comment:
I think it's helpful to have a flag to to set the desired implementation class name to help with this situation without needing to hack with the jar

Reply:
Actually maybe it's fine to crash instead of warn too since it's so dangerous to use a wrong storage.

Reply:
> I think it's helpful to have a flag to to set the desired implementation class name to help with this situation without needing to hack with the jar

So you mean just providing a flag instead of SPI? or both?

Reply:
Both - the flag would be when the SPI is ambiguous but usually there'd just be 0-1 SPI providers and no need for flag.